### PR TITLE
8336874: WhiteBoxAPI: assert(!method->is_abstract() && (osr_bci == InvocationEntryBci || !method->is_native())) failed: cannot compile abstract/native methods

### PR DESCRIPTION
**Pull request by SoniaZaldana**
Hi all, 

This PR addresses [8336874](https://bugs.openjdk.org/browse/JDK-8336874) ensuring enqueuing an abstract method for compilation doesn't hit an assert with WhiteBox. 

Testing: 
- [x] Added test case passes. 

Thanks, 
Sonia

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336874](https://bugs.openjdk.org/browse/JDK-8336874): WhiteBoxAPI: assert(!method-&gt;is_abstract() &amp;&amp; (osr_bci == InvocationEntryBci || !method-&gt;is_native())) failed: cannot compile abstract/native methods (**Bug** - P5)

### Reviewers
 * [Leonid Mesnik](https://openjdk.org/census#lmesnik) (@lmesnik - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20973/head:pull/20973` \
`$ git checkout pull/20973`

Update a local copy of the PR: \
`$ git checkout pull/20973` \
`$ git pull https://git.openjdk.org/jdk.git pull/20973/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20973`

View PR using the GUI difftool: \
`$ git pr show -t 20973`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20973.diff">https://git.openjdk.org/jdk/pull/20973.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20973#issuecomment-2348922494)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/whitebox/TestCompileAbstractMethod.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import compiler.whitebox.CompilerWhiteBoxTest;
+
+public class TestCompileAbstractMethod {
+    private static final WhiteBox WHITE_BOX = WhiteBox.getWhiteBox();
+
+    public static void main(String[] args) throws NoSuchMethodException {
+        Method m1 = A.class.getDeclaredMethod("run");
+        assert m1 != null;
+        WHITE_BOX.enqueueMethodForCompilation(m1, CompilerWhiteBoxTest.COMP_LEVEL_FULL_OPTIMIZATION);
+        if (WHITE_BOX.isMethodCompiled(m1)) {
```
Comment (lmesnik) [inline test/hotspot/jtreg/compiler/whitebox/TestCompileAbstractMethod.java:48]: I think it is better to check that 'enqueueMethodForCompilation' returns false for abstract method additionally.
The isMethodCompiled should always returns true even if WB try to push abstract method but it is rejected on later steps.

**Review by lmesnik — APPROVED** [review]
The overall fix looks good, see small test suggestion in comment.

File: test/hotspot/jtreg/compiler/whitebox/TestCompileAbstractMethod.java (hunk 1 above)
Comment (lmesnik) [reply test/hotspot/jtreg/compiler/whitebox/TestCompileAbstractMethod.java:48]: Sorry, I mistaken. I meant that `isMethodCompiled` should always return false for abstract method. The compiler can't compile it. So this check is fine. But for your fix it is better to check return value of `enqueueMethodForCompilation` as I mentioned.

Comment (SoniaZaldana) [reply test/hotspot/jtreg/compiler/whitebox/TestCompileAbstractMethod.java]: Understood, thanks for checking! I pushed an update checking the return value of `enqueueMethodForCompilation` instead.

**Review by lmesnik — APPROVED** [review]
Thanks for updating test!

**Review by tstuefe — APPROVED** [review]
Good

**Comment by SoniaZaldana** [comment]
Thanks @tstuefe @lmesnik!

**Pull request closed without merging**